### ISSUE TITLE
Flowbite's handling of Turbo is using the wrong strategy

### ISSUE THREAD

**Issue by daniel-rikowski** [themesberg/flowbite#796]
**tl;dr:** The current approach for integrating Turbo with Flowbite is using a mismatched strategy. As far as I can tell, it will only ever work for a certain subset of Turbo-enabled web applications. Events don't work for every case and should be replaced with `MutationObserver`.

--- 
## Details

Sorry for the long wall of text, but I believe this is necessary to understand the problem.

I spend some hours over the weekend to make Flowbite work with Turbo, but hand no actual success. As soon as I used more than just the page cache feature, simple visits or most-of-the-page Turbo frames, I encountered dead ends.

The unpublished fix in https://github.com/themesberg/flowbite/pull/760 did help at first, but it then revealed a huge conceptual problem, which I think cannot be tackled with the current approach.

Let me explain 😃  Most previous issues regarding Turbo were about missing initializing on frame navigations or after stream actions or about broken re-initialization: https://github.com/themesberg/flowbite/issues/450, https://github.com/themesberg/flowbite/pull/760, https://github.com/themesberg/flowbite/issues/698, https://github.com/themesberg/flowbite/issues/697, https://github.com/themesberg/flowbite/issues/454

They are fixed now, but they all are more or less centered around a certain aspect:

**Some component instances need to be initialized or reinitialized after a Turbo operation.**

For a simple application, where Turbo is used mostly as a page cache, this is all you need.

But as soon as you start using Turbo in a more fine-grained approach, with many frames and stream actions, you will also encounter the *opposite* problem:

**Some component instances _must not_ be reinitialized after a Turbo operation.**

There are no reported issues regarding this topic _yet_, besides the odd "no, this fix still doesn't work for me" comments.
(Most likely because fine-grained decomposed Turbo pages utilize stream actions, which didn't work until https://github.com/themesberg/flowbite/pull/760.)

### Example

You have a modal, make it visible, and then update a completely unrelated portion of the DOM, either via frame navigation or
a stream action. Then the subsequent `initFlowbite`/`initModals` will re-initialize *all* modals, even the one currently visible.
This leads to an inconsistent state of the page, because the new instance of `Modal` thinks, it is _hidden_, but the
DOM nodes of the modal still have the TailwindCSS classes, which make it _visible_. The backdrop is removed, too, so you are
left with kind of a "zombie" modal, not fully open, not fully closed.

Depending on the app (or the component) that might not be a problem, e.g. when no Modal is visible during a Turbo operation.
Or when it is actually the modal, which is replaced during that operation, because then you get "fresh" HTML with the correct classes.

In apps where this actually *is* a problem, the idea to make `initFlowbite`/`initModals` *idempotent* (as proposed in https://github.com/themesberg/flowbite/issues/119#issuecomment-1703815359) only would help in those cases, where components are not *replaced*. In other words, in such cases where the DOM nodes of a component get replaced by a Turbo operation, then the component _must_ actually be reinitialized, and an idempotent init function would wrongly skip it.

### Why just Flowbite?

The reason why other frameworks don't have these problems is that their implementation of a component instance manager stores the instances together with the DOM node element, in Foundation indirectly using jQuery's `data`, for example. (See https://github.com/foundation/foundation-sites/blob/develop/js/foundation.core.plugin.js, https://github.com/jquery/jquery/blob/main/src/data/Data.js)

So on those frameworks, if any operation of Turbo replaces a part of the DOM, any affected component instances are gone as well, as opposed to Flowbite, where the DOM nodes and their respective component instances are *loosely* coupled only through an ID string. I like Flowbite's approach better, but it requires a more fine-grained approach, as I explain below.

## The wrong strategy

The core of the problem lies in the event-based approach coupled with a full (re-)initialization after `turbo:load`, `turbo:frame-load` and now the custom `turbo:after-stream-render` event.

Turbo's fine-grained modifications the DOM and the current catch-all approach of (re-)initializing everything on the page conceptually don't match.

## A better strategy

The proper solution is to only (re-)initialize components in those parts of the DOM which *actually* have been changed.

I wrote a simple proof-of-concept handler for Turbo which utilizes MutationObserver and - as far as I can tell - only requires superficial modifications to the current Flowbite JS code:

```js
const observer = new MutationObserver((mutationList, observer) => {
    const modals = window.FlowbiteInstances.getInstances('Modal') // Hack, because `instances` is not exported
    for (const mutation of mutationList) {
        // Destroy components from removed DOM nodes
        for (const root of mutation.removedNodes) {
            const ids = collectIDs(root) // Implemented elsewhere
            const removed = ids.map(id => modals[id]).filter(Boolean)
            removed.forEach(modal => modal.destroyAndRemoveInstance())
        }

        // Initialize components from added DOM nodes
        for (const root of mutation.addedNodes) {
            if (root.nodeType !== Node.ELEMENT_NODE) {
                continue // Most often a #text node
            }
            initModals(root) // <== New parameter required in Flowbite initializers
        }
    }
})

observer.observe(document.body, { childList: true, subtree: true })
```

As you can see, the overall principle is not that complicated, but it requires some modifications to the `init*` functions, most notably adding an optional parameter to specify which subtree to initialize.

The required changes look like this:

```js
export function initModals(root = document) { // <== New optional parameter
    // initiate modal based on data-modal-target
    const targets = Array.from(root.querySelectorAll('[data-modal-target]'))
    // This part is ugly: `querySelectorAll` only considers child nodes and since a non-document node could already
    // be the component target, we must explicitly check the root node, too.
    if (root.matches('[data-modal-target]')) {
        targets.push(root)
    }
    targets.forEach(function ($triggerEl) {
        var modalId = $triggerEl.getAttribute('data-modal-target');
        var $modalEl = document.getElementById(modalId); //
        // ...
    }
    // ...
}
```

I tested this extensively with modals, and it seems to work much better: 

* New nodes/modals are initialized.
* Replaced nodes/modals are destroyed then initialized again.
* Removed nodes/modals are destroyed.
* Unchanged nodes/modals remain untouched.

Most likely, my code can be optimized and perhaps there are some edge cases I didn't consider, especially in the other components I didn't test.

The beauty of this solution is that it is almost completely agnostic to Turbo and should work with every other library which modifies the DOM in a similar fashion.

The only Turbo-specific thing is the event name for the initial page load:

```js
document.addEventListener('turbo:load', () => {
    initFlowbite()
    observeChanges()
})
```

On the other hand, I'm not sure how my solution behaves in conjunction with JS frameworks like Vue or React.
I can imagine might will be conflicts, but this might also make some Vue wrappers obsolete, especially if their only purpose is fine-grained initialization.

But I have some preliminary experience with Stimulus: 
Before coming to this solution, I used a Stimulus controller and manual initialization (i.e. explicitly calling `new Modal`) to tackle the problem described above. This worked, but the observer approach made it obsolete, which is not surprising, since Stimulus also uses `MutationObserver`. It now only provides an "open modal after fame load" feature I wanted, since that feature was removed from Flowbite (https://github.com/themesberg/flowbite/pull/141#issuecomment-1925815453)

As much as https://github.com/themesberg/flowbite/pull/760 was pivotal in finding a proper solution, I must admit that the Turbo developers were right in not adding that event. I once asked for a similar feature in order to apply the same "sledgehammer" approach (https://github.com/hotwired/turbo/pull/425#issuecomment-1246358996 😌 )  

## Final words

Thank you for reading this. I'm hoping my proposal can be integrated into Flowbite. The changes should be small, albeit at numerous locations, i.e. all `init` functions. They should not affect non-Turbo users at all, too, but they would provide _proper_ support for all Turbo-related use cases.

**Comment by Doddlin** [comment]
Could this relate (resolve?) to #793 ?

**Comment by zoltanszogyenyi** [comment]
Hey @daniel-rikowski,

Thanks a lot for the very informative issue and explanations!

I've recently merged a PR that should address this issue:

https://github.com/themesberg/flowbite/commit/3f2ed0acec6767e6ded5a86cbe3651b7ee55f5c4

Can you please have a look at it whether it solves the turbo reload issue?

Otherwise, if you create a PR based on this I will review it and test it and merge it in case it will permanently resolve these turbo reload issues with Flowbite :) you'll get a spot as one of the contributors of Flowbite on the homepage too!

Cheers,
Zoltan

**Comment by opedrosouza** [comment]
I don't know if my issue is related to this, but based on the context I think so.

Basically, I have a `turbo_frame(id: :modal)` on the body of my HTML, and all my CRUD operations are handled inside the modals. 

When I open a modal on the route `app/students` then navigate to the route `app/parents` and go back to the `app/students` route the modal that I have opened before blinks on the screen.

https://github.com/themesberg/flowbite/assets/30217153/79d877ee-9ff6-4759-8929-76971eac3ec9

here is a screen record of the behavior I'm facing.

**Comment by daniel-rikowski** [comment]
Hello @ zoltanszogyenyi, thank you for attending this!

Yes, I'm aware of #760, it filled the last gap on the **"must reinitialize" side** of Turbo, namely after changes caused by stream actions. (Also it enabled me to investigate the problem in the first place.)

But it does not solve the **"must _not_ reinitialize" side** of Turbo, for parts of the page which must not be touched. Even though the approach is very elegant, it funnels even the most minor change of the page through undifferentiated, broad-brushed "reinitialize all the things" `init-*` handlers. (Imagine relevant meme here 😆)

Still, at the moment #760 only makes things _better_. ( :shipit: ?)

I'm tempted to try a PR, although I find the step to delve into a new and foreign codebase to be daunting. Perhaps I'll try to create a small monkey patch first, to quickly aid people in determining if their specific Turbo-related problem is actually _this_ problem.

**Comment by daniel-rikowski** [comment]
Hello @opedrosouza

it might be this problem, but considering the specific timing of the blinking in the screen recording, my first guess would be a "stale cache" or a preview problem.

You can test this quickly, by adding this to all pages, i.e. the layout file:

```html
<head>
  ...
  <meta name="turbo-cache-control" content="no-cache">
</head>
```

If the problem is gone (and your site feels considerably slower :cry:) then you can try to gradually reintroduce Turbo caching, 
first by using `no-preview` instead of `no-cache`. If that works, then by only applying it to certain pages or by manually clearing the cache. (Both possible from JS) Also `[data-turbo-temporary]` may help in your specific case. For details see https://turbo.hotwired.dev/handbook/building#opting-out-of-caching

Perhaps in the future Flowbite can react to `turbo:before-cache` or `[data-turbo-preview]`, perhaps even just by providing some documentation for typical use cases. (Modals via `turbo-frame` seems to be one of those.)

**Comment by opedrosouza** [comment]
> Hello @opedrosouza
> 
> it might be this problem, but considering the specific timing of the blinking in the screen recording, my first guess would be a "stale cache" or a preview problem. (i.e. the Turbo cache, not the browser cache)
> 
> You can test this quickly, by adding this to all pages, i.e. the layout file:
> 
> ```
> <head>
>   ...
>   <meta name="turbo-cache-control" content="no-cache">
> </head>
> ```
> 
> If the problem is gone (and your site feels considerably slower 😢) then you can try to gradually reintroduce Turbo caching, first by using `no-preview` instead of `no-cache`. If that works, then by only applying it to certain pages or by manually clearing the cache. (Both possible from JS) Also `[data-turbo-temporary]` may help in your specific case. For details see https://turbo.hotwired.dev/handbook/building#opting-out-of-caching
> 
> Perhaps in the future Flowbite can react to `turbo:before-cache` or `[data-turbo-preview]`, perhaps even just by providing some documentation for typical use cases. (Modals via `turbo-frame` seems to be one of those.)

@daniel-rikowski it worked, but I think that this is not something truly related to cache itself.

I have a custom js controller which handle the modals on my app

the controller.js file here:
```js
// controller.js
import { Controller } from "@hotwired/stimulus";
import { Modal } from "flowbite";

export default class extends Controller {
  constructor() {
    super(...arguments);
    this.modal = new Modal(this.element);
    this.destroyMissingBackDrop();
  }

  connect() {
    this.modal.show();
  }

  hide() {
    this.element.removeAttribute("src")
    this.modal.destroy();
  }

  // used for form submission as turbo:submit-end->modal--component#submitEnd
  submitEnd(e) {
    if (e.detail.success) {
      this.hide();
    }
  }

  destroyMissingBackDrop() {
    document.querySelector("[modal-backdrop]")?.remove();
  }

}
```
the HTML file here:
```html
<%= tag.turbo_frame(id: :modal) do %>
  <div data-controller="modal--component" tabindex="-1" aria-hidden="true" class="fixed top-0 left-0 right-0 z-50 hidden w-full p-4 overflow-x-hidden overflow-y-auto md:inset-0 h-[calc(100%-1rem)] max-h-full">
    <div class="relative w-full max-w-md max-h-full">
      <!-- Modal content -->
      <div class="relative bg-white rounded-lg shadow dark:bg-gray-700">
        <button type="button" class="absolute top-3 right-2.5 text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm w-8 h-8 ml-auto inline-flex justify-center items-center dark:hover:bg-gray-600 dark:hover:text-white" data-action="click->modal--component#hide">
          <svg class="w-3 h-3" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 14 14">
            <path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"/>
          </svg>
          <span class="sr-only">Close modal</span>
        </button>
        <div class="px-6 py-6 lg:px-8">
          <h3 class="mb-4 text-xl font-medium text-gray-900 dark:text-white"><%= title %></h3>
          <%= body %>
        </div>
      </div>
    </div>
  </div>
<% end %>
```

I had to post the code to say that when the user clicks on the the button to close the modal it works perfectly, but when the user clicks on the backdrop or press `Esc` to close the modal it will behave like the video I posted [here](https://github.com/themesberg/flowbite/issues/796#issuecomment-1931084202)

I tried to find a solution to handle the click on the backdrop element on my js file but without success so far.

**Comment by secretpray** [comment]
> > Hello @opedrosouza
> > it might be this problem, but considering the specific timing of the blinking in the screen recording, my first guess would be a "stale cache" or a preview problem. (i.e. the Turbo cache, not the browser cache)
> > You can test this quickly, by adding this to all pages, i.e. the layout file:
> > ```
> > <head>
> >   ...
> >   <meta name="turbo-cache-control" content="no-cache">
> > </head>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If the problem is gone (and your site feels considerably slower 😢) then you can try to gradually reintroduce Turbo caching, first by using `no-preview` instead of `no-cache`. If that works, then by only applying it to certain pages or by manually clearing the cache. (Both possible from JS) Also `[data-turbo-temporary]` may help in your specific case. For details see https://turbo.hotwired.dev/handbook/building#opting-out-of-caching
> > Perhaps in the future Flowbite can react to `turbo:before-cache` or `[data-turbo-preview]`, perhaps even just by providing some documentation for typical use cases. (Modals via `turbo-frame` seems to be one of those.)
> 
> @daniel-rikowski it worked, but I think that this is not something truly related to cache itself.
> 
> I have a custom js controller which handle the modals on my app
> 
> the controller.js file here:
> 
> ```js
> // controller.js
> import { Controller } from "@hotwired/stimulus";
> import { Modal } from "flowbite";
> 
> export default class extends Controller {
>   constructor() {
>     super(...arguments);
>     this.modal = new Modal(this.element);
>     this.destroyMissingBackDrop();
>   }
> 
>   connect() {
>     this.modal.show();
>   }
> 
>   hide() {
>     this.element.removeAttribute("src")
>     this.modal.destroy();
>   }
> 
>   // used for form submission as turbo:submit-end->modal--component#submitEnd
>   submitEnd(e) {
>     if (e.detail.success) {
>       this.hide();
>     }
>   }
> 
>   destroyMissingBackDrop() {
>     document.querySelector("[modal-backdrop]")?.remove();
>   }
> 
> }
> ```
> 
> the HTML file here:
> 
> ```
> <%= tag.turbo_frame(id: :modal) do %>
>   <div data-controller="modal--component" tabindex="-1" aria-hidden="true" class="fixed top-0 left-0 right-0 z-50 hidden w-full p-4 overflow-x-hidden overflow-y-auto md:inset-0 h-[calc(100%-1rem)] max-h-full">
>     <div class="relative w-full max-w-md max-h-full">
>       
>       <div class="relative bg-white rounded-lg shadow dark:bg-gray-700">
>         <button type="button" class="absolute top-3 right-2.5 text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm w-8 h-8 ml-auto inline-flex justify-center items-center dark:hover:bg-gray-600 dark:hover:text-white" data-action="click->modal--component#hide">
>           <svg class="w-3 h-3" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 14 14">
>             <path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"/>
>           </svg>
>           <span class="sr-only">Close modal</span>
>         </button>
>         <div class="px-6 py-6 lg:px-8">
>           <h3 class="mb-4 text-xl font-medium text-gray-900 dark:text-white"><%= title %></h3>
>           <%= body %>
>         </div>
>       </div>
>     </div>
>   </div>
> <% end %>
> ```
> 
> I had to post the code to say that when the user clicks on the the button to close the modal it works perfectly, but when the user clicks on the backdrop or press `Esc` to close the modal it will behave like the video I posted [here](https://github.com/themesberg/flowbite/issues/796#issuecomment-1931084202)
> 
> I tried to find a solution to handle the click on the backdrop element on my js file but without success so far.

I had the same problem. I solved it this way 

```js
import { Controller } from "@hotwired/stimulus"

// Connects to data-controller="modals"
export default class extends Controller {
  static values = { 
    modalId: { type: String, default: "modal"},
  }

  initialize() {
    this.#addCacheControlMetaTag()
  }

  connect() {
    this.showModal()
  }

  disconnect() {
    this.destroyModal()
    this.#removeCacheControlMetaTag()
  }

  closeModalOnEscape(event) {
    if (event.key === "Escape") {
      this.closeModal(event);
    }
  }

  showModal() {
    this.element.classList.remove("hidden")
    this.element.classList.add("flex")
    this.modalBackdrop.classList.remove("hidden")
    this.bodyElement.classList.add("overflow-hidden")
  }

  destroyModal() {
    if (this.element) {
      this.element.classList.add("hidden")
      this.element.classList.remove("flex")
    }
    if (this.modalBackdrop) this.modalBackdrop.classList.add("hidden")
    this.bodyElement.classList.remove("overflow-hidden")
    
    if (this.turboFrame) {
      this.turboFrame.src = ""
      this.turboFrame.removeAttribute("complete")    
    }
  }

  closeModal(event) {
    event.preventDefault()
    this.destroyModal() 
  }

  disconnect() {
    this.destroyModal()
  }

  get modalBackdrop() {
    return document.querySelector("#modalBackdrop")
  }

  get bodyElement() {
    return document.querySelector("body")
  }

  get turboFrame() {
    return document.querySelector(`turbo-frame#${this.modalIdValue}`)
  }

  #addCacheControlMetaTag() {
    const meta = document.querySelector('meta[name="turbo-cache-control"]')

    if (!meta) {
      Turbo.cache.exemptPageFromCache()
      // or
      // const meta = document.createElement('meta')
      // meta.name = 'turbo-cache-control'
      // meta.content = 'no-cache'
      // document.head.appendChild(meta)
    }
  }

  #removeCacheControlMetaTag() {
    document.querySelectorAll('meta[name="turbo-cache-control"][content="no-cache"]').forEach(el => el.remove());
  }
}
```

**Comment by tbreier** [comment]
I came across an issue during Turbo Stream replace responses with reproduction steps which _might_ be related: #932

**Comment by UnconventionalMindset** [comment]
Is there something more recent than https://github.com/themesberg/flowbite/pull/807 on the works?

**Comment by daniel-rikowski** [comment]
The solution I described, or better said - the concept itself - works quite well. I'm using it in production for an in-house app.

Since my PR was not merged, and I didn't want to maintain a fork, I switched to the plain JS API of Flowbite and use Stimulus to create and destroy the instances, while the HTML is encapsulated in ViewComponents. 

There were some edge cases, like morphing, but eventuelly this worked very well. Unfortunetly, most of my code is tailored to my app. Without some work, I think it's of limited use to others.

In a nutshell: If you need a working solution right now, use Stimulus to "listen" for HTML changes - wherever they might come from Turbo/Cache/Frames/Streams - and then create the instances by hand. (Two controllers per trigger/target pair)

**Comment by jondkinney** [comment]
@daniel-rikowski that's some helpful direction, thanks for all the work to track this down! We're experiencing some of these same problems in our app now, too. 

I'm a long time Rails dev, but am a bit newer to Stimulus and Turbo. I get the high-level concept, and I've used MutationObservers before lightly, would you be willing to to post a generic-ish example of how you've solved this? And am I understanding correctly that you're avoiding the "turbo" variant of the flowbite.js when coupled with this more manual solution?

**Comment by jondkinney** [comment]
@daniel-rikowski I'm in the middle of the implementation as you describe (I have ditched the turbo version of the flowbite.js) and have a stimulus controller that uses a `mutationObserver` to watch for changes in a list of tasks in our system. If one of those task nodes is new, we initialize a flowbite drawer for it, so that when clicking the new task in the list that was put there via a turbo_stream, it has a drawer that will slide in from the right. This is working great.

However, none of the data-attribute driven links/buttons within that drawer work after manual initialization. For example, the `x` button in the upper right of the drawer that should close the drawer that its contained within, doesn't fire, despite the fact that the `data-drawer-target` that it's pointing to exists because it was just initialized. Do you have to re-bind all the individual click events for those data-attr driven drawer elements when you initialize a new drawer like that? 

In addition to the `x` close button that isn't working in the new drawer, we have a link to hide the current drawer and show its parent (they swap with each other, essentially). But that doesn't work, either. 

Maybe there's a generic way to do that for any data-attributes associated with the new data-drawer-target?

Perhaps I'm missing something in your described solution though, too. I'm not sure I fully understand the "two controllers per trigger/target pair" paradigm that you mentioned just a few weeks ago, but if feels related. Is it that I'm currently only handling the _target_ initialization and I need another controller for all the _triggers_? Rubber ducking here, that seems right. But can you confirm? Also if you have any battle tested code for the reinitialization of the trigger side of things you could share, that seems like it'd be pretty generic. Thanks!

**Comment by jondkinney** [comment]
@daniel-rikowski this is really great, I can't thank you enough! I'm following everything you laid out, and I'm in the process of implementing it for the Drawer now. I see that the trigger and target drawer controllers are inheriting from the other more generic parent classes (controllers) and I understand how I'll be able to replicate that pattern for the other components like modals and dropdowns, etc. 

That said, I do have a few follow up questions: 

1. Do you use this pattern for _every_ component that you use from Flowbite? I'm guessing so, but wanted to confirm.
2. Is the stimulus controller's connect method and resulting initialization the way you initialize all of your Flowbite components _on initial page load_ as well as when streaming them in from a turbo_stream via a partial render from the controller? If so, how do you prevent the regular flowbite.js from also initializing them on page load? 
3. How are you including the Flowbite JS in your app? I understand that you're not using the turbo variant, but are you still using the minified version from the `importmap.rb` via CDN? Or are you using `importmap.rb` but vendoring it? The way you're importing Drawer in the example components makes me thing you're doing it different than me/what they have documented, and I'm curious what that approach is since I'm guessing I'll need to follow the same pattern to have this all work.

```js
import { default as Drawer } from 'flowbite/lib/esm/components/drawer'
```

That looks to imply that you have flowbite installed in node_modules and you're importing it directly, but I can't get that import path to work with how I'm including flowbite, itself 🤔 I tried `importmap.rb` with the minified version via cdn, and the esm version when vendoring, and also tried pinning it to the npm package via yarn, but nothing I did allowed me to import the Drawer module like you show. I can, however, "just use" the Drawer global directly? That seems...not ideal? But I'm not 100% sure. All of the latest asset changes in Rails have me a bit confused.

**Comment by daniel-rikowski** [comment]
I'm glad my code helped you 😊 

Answering your questions:

1. Yes, I applied that pattern for every Flowbite component. In addition I have more specialized controllers which add functionality to the HTML elements, for example a dropdown for a search field: That dropdown has the Flowbite dropdown controller and a controller which fetches proposals. (i.e. "composition" of multiple controllers)

2. I don't use flowbite.js and its initialization logic. Instead I import the components directly and perform all initialization using my Stimulus controllers. 

3. I don't know how to translate the import statement to work with import maps, somehow I always hit a wall when I tried them. I'm using webpack and lately vite. My current "asset stack" is vite with vite-rails, tailwindcss, yarn with pnp, view components.

The `import ... from 'flowbite/lib/` statements are my only imports from JS. All CSS importing is done via `tailwind.config.js` and the Flowbite plugin.

**Comment by daniel-rikowski** [comment]
Updated the gist to include the remaing controllers (except for the placeholders)

**Comment by jondkinney** [comment]
Thanks @daniel-rikowski! I'm working on implementing this now with importmaps. I think I have a path I can follow to make it work. I'll let you know if I crack that nut. One question about the directory structure that you're using, and I get that this is probably just a preference thing, I'm just curious how you have it setup - are you nesting the `flowbite/` folder in your `app/javascript/controllers` folder so it'd be `app/javascripts/controllers/flowbite/drawer_trigger_controller.js`, for example? Or are you putting the `flowbite/` folder next to the `controllers/` folder in your hierarchy? In which case it'd be `app/javascripts/flowbite/drawer_trigger_controller.js`. I'm wondering because it has implications for the stimulus eager loading controller config, and I didn't know if you'd already solved some edges there or not. 

My thought is that the flowbite folder is next to the controllers folder, in which case the `app/javascripts/controllers/index.js` file would look something like this:

```js
// Import and register all your controllers from the importmap via controllers/**/*_controller
import { application } from "controllers/application"
import { eagerLoadControllersFrom } from "@hotwired/stimulus-loading"
eagerLoadControllersFrom("controllers", application)
eagerLoadControllersFrom("../flowbite", application)
```

If you nest the `flowbite/` folder inside the `app/javascripts/controllers/` folder maybe the `eagerLoadControllersFrom` would cover the subdirectories automatically? It didn't seem to work that way when I tried it. But I could have had something misconfigured.

**Comment by daniel-rikowski** [comment]
This is my directory structure (incomplete, of course)


```
📁 app
  📁 frontend ("javascripts")
    📁 channels (ActionCable)
      📄 index.js
    📁 components (ViewComponents, it's debatable if this should instead be in `app/views`...)
      📁 ...
    📁 controllers
      📁 flowbite
        📄 ...
        📄 trigger_target.js
        📄 ...
      📄 ...
      📄 other_controller.js
      📄 ...
    📁 entrypoints
      📄 application.js (=main entry point)
    📄 application.css
    📄 stimulus.js
    📄 ...
```

I'm not using `eagerLoadControllersFrom` because I wasn't aware of it (🤷 ) But glancing over the source code, I guess this is tailored to importmaps (not sure...) It looks like it doesn't work recursively, judging from [this line](https://github.com/hotwired/stimulus-rails/blob/f5c7475abf8701ebb96cecb7e99c6b5bf876452f/app/assets/javascripts/stimulus-loading.js#L8)

Anyway, I'm using Vite's `import.meta.glob` and previously I have used webpack's `require.context`

Like this:

**vite(-rails)**
```js
// Generic controllers
const genericControllers = import.meta.glob( // <== import.meta.glob provided by Vite
    "./controllers/**/*_controller.js", { eager: true, import: 'default' }
)

for (const path in genericControllers) {
    const module = genericControllers[path]
    const name = path
        .match(/controllers\/(.+)_controller\.js$/)[1]
        .replaceAll("/", "--")
        .replaceAll("_", "-")
    application.register(name, module)
}

// Controllers from components
const controllers = import.meta.glob(
    "./components/**/controller.js",
    { eager: true, import: 'default' }
)

for (const path in controllers) {
    const module = controllers[path]
    const name = path
        .match(/components\/(.+)\/controller\.js$/)[1]
        .replaceAll("/", "--")
        .replaceAll("_", "-")
    application.register(name, module)
}
```


**webpack** (not tested)

```js
import { Application } from '@hotwired/stimulus'
import { definitionsFromContext } from '@hotwired/stimulus-webpack-helpers'

const application = Application.start()
const context = require.context('controllers', true, /\.js$/)  // <== require.context provided by webpack
application.load(definitionsFromContext(context))
```

I guess it's not to bad to sprinkle a few `index.js` into the javascript directories, but in my case, where I have many ViewComponents, which many having their own separate `controller.js`, this annoyed me very quickly, so I just went for the "brute force" glob import, which unfortunately is preprocessor-specific...

**Comment by jondkinney** [comment]
Nice, thanks for the primer on your setup @daniel-rikowski. I'm trying to avoid a bundler, and I was able to get everything working with importmaps! It was quite a journey to understand all of the pieces, and at one point I was 99% sure I was going to have to vendor the flowbite components and modify their own import lines to manually pinned versions of their dependencies like `flowbite/dom/instances` and `@popperjs/core` since the component source files were using relative file based imports in their node module resolution style. 

That _did_ work, and even though it was only 1-2 lines that needed to be modified per flowbite component, I wasn't nuts about having to maintain custom versions of Drawer, Modal, etc. in my vendor/javascript files and potentially have to update them on each release. Not to mention, the manual pinning of not just the components themselves, but also all of their dependencies muddying up my `importmap.rb` file.

What I landed on is pretty elegant! Through trial and error and some sessions with Claude Sonnet 3.7, I discovered https://www.jsdelivr.com/esm which touts itself as

> A New-Age CDN for JavaScript modules. Load modern JavaScript packages built for you on-demand. Works in modern web browsers, node.js, and deno

I'd come across the CDN option with importmaps in the past and have had various of them at one point or another. Here are some examples, none of which I'm actually using in my app, but they illustrate that there are several of these CDNs around.

- https://www.jsdelivr.com (the non-esm variant of the same root service)
- https://cdn.jsdelivr.net/npm/vue@2.6.14/dist/vue.esm.browser.js
- https://ga.jspm.io/npm:jquery@3.7.0/dist/jquery.js

...but until finding the **`esm.run`** variant, I'm not sure that any of the other options resolve all of the relative path based import dependencies like it does with this setup. Here's how it works - 

```ruby
# importmap.rb

# ... other code

flowbite_components = %w[
  accordion carousel clipboard collapse datepicker dial dismiss drawer dropdown input-counter modal popover tabs tooltip
]
flowbite_components.each do |component|
  pin "flowbite/components/#{component}", to: "https://esm.run/flowbite@2.5.2/lib/esm/components/#{component}/index.js"
end
```

Perhaps obviously, this loop results in individual pins for each component

```ruby
pin "flowbite/components/modal", to: "https://esm.run/flowbite@2.5.2/lib/esm/components/modal/index.js"
```

This shows up in the browser like this (only pinning modal in this example)

![Image](https://github.com/user-attachments/assets/f9dc9a8c-0731-47ab-86e7-78ddbeb412f0)

Which then allows for importing each of those modules in their component specific _trigger and _target controllers, `modal_trigger_controller.js`, for example.

```js
import Modal from 'flowbite/components/modal'
```

You also have to register and pin the controllers (if you're already using an importmap setup this is probably already done), But that's where the `eagerLoadControllersFrom("controllers", application)` from `app/javascript/controllers/index` and `pin_all_from "app/javascript/controllers", under: "controllers"` from `importmap.rb` come into play. You also need to `import "controllers"` in `app/javascript/application.js` to load them in the app. 

`eagerLoadControllersFrom` is recursive and will register all `_controller.js` files in all subdirectories of `app/javascript/controllers`. Then `pin_from_all` is also recursive, as is the import statement, so everything should be locked and loaded

So if we check our importmap setup by running `bin/importmap json` from the root of the project, we'll see the json object the browser uses to figure out which url to resolve to when importing one of the pin aliases (object truncated to highlight the important parts). You can also see this object in your `head` tag, but running it at the command line is faster for iterating while developing. Sort of like running `rake routes`.

```json
{
  "imports": {
    "application": "/assets/application-1934cfff.js",
    "@hotwired/stimulus": "/assets/stimulus.min-4b1e420e.js",
    "@hotwired/stimulus-loading": "/assets/stimulus-loading-1fc53fe7.js",
    "@hotwired/turbo-rails": "/assets/turbo.min-1262d735.js",
    "flowbite/components/modal": "https://esm.run/flowbite@2.5.2/lib/esm/components/modal/index.js",
    "controllers/application": "/assets/controllers/application-a5438139.js",
    "controllers/flowbite/modal_target_controller": "/assets/controllers/flowbite/modal_target_controller-d45f1f35.js",
    "controllers/flowbite/modal_trigger_controller": "/assets/controllers/flowbite/modal_trigger_controller-c3a62f59.js",
    "controllers/flowbite/target_controller": "/assets/controllers/flowbite/target_controller-06a7058b.js",
    "controllers/flowbite/trigger_controller": "/assets/controllers/flowbite/trigger_controller-13776067.js",
    "controllers/mixins/use_morph_events": "/assets/controllers/mixins/use_morph_events-5b95a5d3.js",
  }
}
```

Which means we have access to import both our `TriggerController` and `Modal` in context like so where `controllers/flowbite/trigger_controller` and `flowbite/components/modal` are our registered pins that will resolve to the URL specified in the importmap.

```js
import { default as TriggerController } from 'controllers/flowbite/trigger_controller'
import Modal from 'flowbite/components/modal'

export default class extends TriggerController {
  #previousFocus = null

  static outlets = ['flowbite--modal-target']
  static values = {
    backdrop: { type: String, default: 'dynamic' }, // 'dynamic' or 'static'
    backdropClasses: { type: String, default: 'bg-gray-900/50 dark:bg-gray-900/80 fixed inset-0 z-40' },
    closable: { type: Boolean, default: true },
    placement: { type: String, default: 'top-center' }, // {top|center|bottom}-{left|center|right} (eg. top-left or bottom-right)
    action: { type: String, default: 'toggle' }, // 'toggle', 'show', 'hide'
  }

  connect() {
    console.log("flowbite-modal-trigger connected")

    super.connect()
    this.onClick = this.click.bind(this)
    this.element.addEventListener('click', this.onClick)
  }

  // ... rest of the modal initialization code
}
```

What's really cool, is that the CDN's ESM infrastructure will also allow for pinned modules importing separate packages that it will automatically download and make work through updates to the bundled module, as necessary. Here we show that the `datepicker` flowbite component requires a separate `flowbite-datepicker@1.3.0` package, and jsdeliver's esm import resolution makes it "just work".

![Image](https://github.com/user-attachments/assets/2a58c438-0dcc-4e73-9e0b-4bfc5cc433d0)

The last piece I didn't understand at first, having never used stimulus outlets previously, was how to invoke any of this from the trigger's data attributes, but I figured it out. Here's an example if anyone else is attempting to use this setup:

```erb
<div
  data-controller="flowbite--modal-target"
  id="flowbite--modal-target"
  class="fixed inset-0 z-50 hidden overflow-y-auto"
  aria-labelledby="modal-title"
  aria-modal="true"
>
  <div class="flex px-4 pb-20 pt-4 sm:block sm:p-0">
    <div class="inline-block transform rounded-lg bg-white px-4 pb-4 pt-5 shadow-xl transition-all sm:my-8 sm:w-full sm:p-6">
      <div class="absolute right-0 top-0 hidden pr-4 pt-4 sm:block">
        <button
          type="button"
          class="rounded-md bg-white text-gray-400 hover:text-gray-500 focus:outline-none"
          data-controller="flowbite--modal-trigger"
          data-flowbite--modal-trigger-flowbite--modal-target-outlet="#flowbite--modal-target"
        >
          <span class="sr-only">Close</span>
          <svg class="h-6 w-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
            <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M6 18L18 6M6 6l12 12"/>
          </svg>
        </button>
      </div>
    </div>
    <!-- modal content here -->
  </div>
</div>
```

In the above modal partial you can see the modal's data-controller target is specified, then the close button specifies the trigger data-controller, and the outlet is the complicated final data attribute in the button code. The semantics of that are `data-[controller-name]-[outlet-name]-outlet="[selector]"`

You can of course trigger the opening of the modal (or whatever other component) from outside the component with any other trigger element

```erb
<button
  type="button"
  data-controller="flowbite--modal-trigger"
  data-flowbite--modal-trigger-flowbite--modal-target-outlet="#flowbite--modal-target"
  class="p-2 md:p-2.5 rounded-lg bg-brand-gold-500 px-3 md:px-4 text-sm text-brand-navy-500 whitespace-nowrap"
>
  Open Modal
</button>
```

One note, I'm omitting the `data-flowbite--modal-trigger-action-value="open"` (or `close` or `toggle`) here because in the modal trigger controller, if no `-action-value` is specified, it defaults to `toggle`. But you could be more explicit if you wanted to. 

```js
// from the modal trigger controller's static values array:
action: { type: String, default: 'toggle' }, // 'toggle', 'show', 'hide'
```

After I got this far, things were looking quite good! BUT there was one major issue left. Closing the modal was acting upon an independent instance of the `Modal` object since every trigger controller was instantiating another `Modal` object. I was able to modify the example `createInstance` methods you provided @daniel-rikowski to early return if they could connect to their outlet's existing instance. This early return avoids the need to create something like a "modal registry" to track existing instances. I was especially keen on finding a way to avoid overcomplicating the "re-use existing component instances" bit because I didn't want to have to have a registry for each component type or a more complicated global one. This works a treat!

There's also a nuance in the click handler below where we need to act upon the _target_ instance rather than the trigger instance. This is related to the early return fix in that we're reaching over to the target instance, but it's technically a separate issue.

```js
createInstance() {
  // If we already have an instance in the target controller, reuse it
  if (this.targetController.instance) {
    console.log("GOOD NEWS EVERYONE! Reusing existing modal instance...")
    this.instance = this.targetController.instance
    return this.instance
  }

  const options = {
    backdrop: this.backdropValue,
    backdropClasses: this.backdropClassesValue,
    closable: this.closableValue,
    placement: this.placementValue,
    // Register direct callbacks that dispatch events to both the trigger and target elements
    onHide: () => {
      console.log("Direct onHide callback fired")
      this.#handleHide()
    },
    onShow: () => {
      console.log("Direct onShow callback fired")
      this.#handleShow()
    },
    onToggle: () => {
      console.log("Direct onToggle callback fired")
      this.#handleToggle()
    }
  }

  const instanceOptions = {
    // id: this.targetElement.id,
    // override: true
  };

  console.log("creating modal instance")
  this.instance = new Modal(this.targetElement, options, instanceOptions)
  console.log("modal instance created", this.instance)

  return this.instance;
}
```

Here's a full modal trigger example

```js
import { default as TriggerController } from 'controllers/flowbite/trigger_controller'
import Modal from 'flowbite/components/modal'

export default class extends TriggerController {
  #previousFocus = null

  static outlets = ['flowbite--modal-target']
  static values = {
    backdrop: { type: String, default: 'dynamic' }, // 'dynamic' or 'static' - dynamic allows the modal to be closed by clicking the backdrop, while static does not
    backdropClasses: { type: String, default: 'bg-gray-900/50 dark:bg-gray-900/80 fixed inset-0 z-40' },
    closable: { type: Boolean, default: true },
    placement: { type: String, default: 'top-center' }, // {top|center|bottom}-{left|center|right} (eg. top-left or bottom-right).
    action: { type: String, default: 'toggle' }, // 'toggle', 'show', 'hide'
  }

  connect() {
    console.log("flowbite-modal-trigger connected")

    super.connect()
    this.onClick = this.click.bind(this)
    this.element.addEventListener('click', this.onClick)
  }

  disconnect() {
    this.element.removeEventListener('click', this.onClick)
    super.disconnect()
  }

  flowbiteModalTargetOutletConnected(outlet, element) {
    this.targetElement = element
    console.log("flowbite-modal-target outlet connected", this.targetElement)
  }

  flowbiteModalTargetOutletDisconnected(outlet, element) {
    this.targetElement = null
    console.log("flowbite-modal-target outlet disconnected")
  }

  #handleToggle(event) {
    // console.log('Modal toggle event received', event)
    if (this.targetController.instance.isVisible()) {
      this.#handleHide(event)
    } else {
      this.#handleShow(event)
    }
  }

  #handleShow(event) {
    // Store the element that had focus before the modal opened
    this.previousFocus = document.activeElement
    // console.log('Modal show event received, previous focus was:', this.previousFocus)

    // Set focus to the first focusable element in the modal (not on the trigger element)
    setTimeout(() => {
      // Important change: Use targetElement instead of this.element
      const focusable = this.targetElement.querySelectorAll(
        'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'
      )
      if (focusable.length > 0) {
        focusable[0].focus()
      }
    }, 50)
  }

  #handleHide(event) {
    // Return focus to the element that was focused before the modal opened
    if (this.previousFocus && typeof this.previousFocus.focus === 'function') {
      console.log('Hide event received, returning focus to:', this.previousFocus)
      this.previousFocus.focus()
    }
  }

  createInstance() {
    // If we already have an instance in the target controller, reuse it
    if (this.targetController.instance) {
      console.log("GOOD NEWS EVERYONE! Reusing existing modal instance...")
      this.instance = this.targetController.instance
      return this.instance
    }

    const options = {
      backdrop: this.backdropValue,
      backdropClasses: this.backdropClassesValue,
      closable: this.closableValue,
      placement: this.placementValue,
      // Register direct callbacks that dispatch events to both the trigger and target elements
      onHide: () => {
        console.log("Direct onHide callback fired")
        this.#handleHide()
      },
      onShow: () => {
        console.log("Direct onShow callback fired")
        this.#handleShow()
      },
      onToggle: () => {
        console.log("Direct onToggle callback fired")
        this.#handleToggle()
      }
    }

    const instanceOptions = {
      // id: this.targetElement.id,
      // override: true
    };

    console.log("creating modal instance")
    this.instance = new Modal(this.targetElement, options, instanceOptions)
    console.log("modal instance created", this.instance)

    return this.instance;
  }

  click(e) {
    // console.log("flowbite-modal-trigger click", e)

    // Use the shared instance set in the targetController, otherwise we'd be
    // calling .toggle(), .show(), .hide() on a new instance of Modal for each
    // trigger if we used this.instance.hide() (or toggle or show)
    const targetInstance = this.targetController.instance

    console.log("flowbite-modal-trigger actionValue", this.actionValue)
    switch (this.actionValue) {
      case 'toggle':
        targetInstance.toggle()
        break
      case 'show':
        targetInstance.show()
        break
      case 'hide':
        targetInstance.hide()
        break
    }
  }
}
```

I think this is most of what someone would need to replicate this setup with importmaps and expand upon your code samples slightly by reusing existing instances @daniel-rikowski. I definitely wouldn't have gotten this working without your code, so thanks so much again. I'm pretty happy with how this is all working! Much more control over the code and experience without _too_ much more overhead (once you understand all the moving pieces). That said, it would be nice to have this all handled automagically @zoltanszogyenyi. The sledgehammer re-init all components on every turbo event approach is a pretty big blocker to anyone wanting to adopt this in a Turbo Rails context.

**Comment by jondkinney** [comment]
@daniel-rikowski one thing I'm running in to now is if I have triggers that are inside of a component / partial that is getting updated via turbo_steams, when one of those outlet disconnects it breaks the containing component instance and it doesn't automatically auto-recover when those outlets re-connect after the update. Here's a more concrete example. 

In our system we have parent tasks and subtasks, I have the parent task drawer open, and I add a new subtask. If I re-render the contents of the parent task drawer to get things updated, even the close button for the parent task drawer component disconnecting (and subsequently re-connecting) causes that orphan issue you mention with the backdrop gone and the parent task drawer somewhat zombified. 

Now, the newly streamed in close button actually _does_ work to close the drawer, I think because when it re-connects it's able to initialize properly, but the open drawer's background is gone.

I've tracked it down to the code I added to re-use the existing instances rather than initializing a new Drawer instance for every trigger that gets created, because I mention above, when doing it that way, the close button (or any other trigger) tries to act upon it's own new Drawer instance rather than the first one that was created. If I remove the code to early return the existing instance then the background doesn't break because a new Drawer is properly initialized after being added to the dom, but then I'm back to the triggers not acting upon the proper Drawer instance from before (and you get double backdrops and other weird breakage).

But maybe I'm missing something fundamental because I'm not sure how you _wouldn't_ have run across that same problem of a close button not closing the drawer because it was acting upon a different instance. 🤔

**Comment by jondkinney** [comment]
@daniel-rikowski I'm still struggling a little bit on components breaking and not reestablishing connection properly. I see that you have `this.instance.destroyAndRemoveInstance()` defined in your `trigger_controller.js` but that method isn't actually defined anywhere. Maybe this is part of the reason I needed to reach for a component registry? The instances aren't being fully destroyed and then can't be re-established? We're seeing breakage of components when a new trigger for an existing component streams into the DOM. But you didn't have a component registry to look up existing instances, so I'm thinking some part of the bit that links the trigger to the target is missing in your code (and my current implementation). 

Is it true that if a new trigger streams into the DOM it should re-initialize a new component that will bind it to all other triggers and targets of the same id specified in the outlet? If so, I'm not seeing that currently and that's the missing piece, but I haven't been able to solve that just yet. Sorry to keep bothering, but if you have any thoughts I'd be grateful! I have a repo I'm working on that'll showcase this whole setup that you could take a look at, too, if you like. https://github.com/headwayio/flowbite_importmaps_stimulus

**Comment by daniel-rikowski** [comment]
`this.instance` contains the Flowbite component, e.g. a `Drawer` instance. The method `destroyAndRemoveInstance` is then defined by Flowbite, e.g. for the Drawer: https://github.com/themesberg/flowbite/blob/e6e19b4e7ea762b2c6727f54b57fd6efe12b4a5b/src/components/drawer/index.ts#L99-L101

A component registry is exactly what Flowbite uses internally, `destroyAndRemoveInstance` calls `removeInstance` which removes the component instance from the instances registry. 

The instances are added by the constructor, e.g.:

https://github.com/themesberg/flowbite/blob/e6e19b4e7ea762b2c6727f54b57fd6efe12b4a5b/src/components/drawer/index.ts#L47-L52

But I'm not using it, frankly I never paid much attention to it, it just seemed like good pratice to clean it up after calling `destroy`.

> Is it true that if a new trigger streams into the DOM it should re-initialize a new component that will bind it to all other triggers and targets of the same id specified in the outlet?

That is my understanding, but there is no real **re-** initialization, because when the new trigger appears, the old one is already destroyed and removed (or at least it should be) 

Do you by any chance have multiple triggers for the same target? This should be supported by Flowbite, but I did not test it with my implementation. Perhaps this is the cause of the problems...?

**Comment by jondkinney** [comment]
Yes, many multiple triggers for the same target. For example, a button (trigger) to open a modal, then a close button `X` in the upper right of the modal, itself, as well as sometimes a "cancel" button in the lower left, too. They all use the trigger/target/outlet system to do their work on the same target. It's the triggers that are _inside_ of the modal/drawer, itself, that cause the breakage. I've been able to avoid the breakage sometimes by using `turbo_stream.update` instead of `.replace` as well as using `method: :morph` on either/both of the `.update`/`.replace` calls, depending on the situation. But I need to find a way to ensure that no matter when a new trigger streams into the DOM, even if it's inside of the target it needs to affect, that it can find its target and work. The component registry that I created allows this to happen, usually...but there are edges, it seems. But without the component registry, clicking the `x` target inside a modal to close it breaks pretty spectacularly.

**Comment by daniel-rikowski** [comment]
I see...

I did indeed stumbled upon this in the beginning, I remember now.

Judging from the code of Flowbite, I don't think this is possible because of the way Flowbite is implemented. Look at this code for initializing the drawer, which is happening for every new instance:

https://github.com/themesberg/flowbite/blob/e6e19b4e7ea762b2c6727f54b57fd6efe12b4a5b/src/components/drawer/index.ts#L55-L81

This directly manipulates the DOM element (`this._targetEl`), changes classes, adds event handlers, etc. and there doesn't seem to be any mechanism for idempotency.

I did not solve this. Instead for the common case of close buttons (`X`) I do something like this: 

1. Find the closest target element using `targetElement = event.target.closest(...)`
2. Get the Stimulus target controller from that element using `targetController = this.application.getControllerForElementAndIdentifier(targetElement, "flowbite--drawer-trigger")`
3. Use the Flowbite `instance` from that controller: `targetController.instance.hide()`

So from some perspective I am indeed using a (convoluted) registry: The DOM 😆

Perhaps a lightweight "close-only" controller might be a solution for you.

**Comment by pldavid2** [comment]
Hello, did notice a similar behaviour. 

Given the issue is still open i understand it was still not possible to be solved from the flowbite side right, and i should follow @jondkinney or @daniel-rikowski solutions, right?

Sorry for the question, rather new with the stimulus/turbo/flowbite combination.

**Comment by pldavid2** [comment]
@zoltanszogyenyi did you see any particular problem with the solution proposed by @daniel-rikowski on first post? 
I'm tempted to vendor flowbite updating the code myself with his suggestion, but maybe you saw some issue and that's why was not added. Currently using the last flowbite / tailwind v4 version.
BTW thanks all for the insights here.

**Comment by artinboghosian** [comment]
@zoltanszogyenyi been playing around with this idea and the fundamental flaw in my approach is that the element is replaced when turbo stream response is returned. And so without reinitializing, the modal is pointing to an element no longer on the page 😦

**Comment by artinboghosian** [comment]
@zoltanszogyenyi Actually I was wrong, I did get this working with the following changes:

```
  const originalAddInstance = window.FlowbiteInstances.addInstance;

  window.FlowbiteInstances.addInstance = function(component, instance, id, override = false) {
    let finalOverride = override;

    if (id) {
      const element = document.getElementById(id);

      if (element?.dataset.flowbiteOverride !== undefined) {
        debugger
        finalOverride = element.dataset.flowbiteOverride === 'true';
      }
    }

    originalAddInstance.call(this, component, instance, id, finalOverride);
  }
```

```
<div id="my-modal" data-flowbite-override="false"></div>
```

The issue previously was that the Turbo Stream response from server was re-rendering the modal each time. This meant that it was originally initialized with a $_targetEl that no longer existed in the DOM (b/c I forced override to "false").

**Comment by daniel-rikowski** [comment]
> The only thing that currently doesn't work is reinitialization of components after a form submit with a 422 error.

I didn't have problems with 422 errors specificially. The problems I noticed arise mostly from the fact, that certain components have two parts, a trigger and a target, and some mutation of the page removes one of them and the other is left "dangling".

I tailored my app around this, by using the Stimulus controllers and outlets: Whenever one part of the component, is removed, it notifies the other one, so it can clean up. Or re-create itself as soon as the new counterpart re-appears. But this code still makes certain assumptions about the page, so it is not a universal solution.

One aspect of the problem is that the Flowbite component initializers assume a clean, uninitialized HTML page. They don't check for potential remnants of previously initialized components, so the often-attempted solution of reinitializing the page doesn't help with this problem.

The next time I'll be working with a new Rails + Flowbite project, I'll attempt to tackle this problem again, but at the moment I have no (paying) Rails projects, so I have to postpone this...

**Comment by mwlang** [comment]
@zoltanszogyenyi @roberttanislav — I want to do a focused piece of work to make Flowbite **first-class for Rails / Hotwire / Stimulus**, addressing this issue and the cluster of related ones in one coordinated push. That means:

**Upstream (this repo, via 3 sequential PRs):**

1. **Idempotency guards** on every `initX()` — re-running becomes a no-op when instances already exist; trigger-binding loops dedupe via owner-aware comparison. Closes #1042 standalone. Zero API change.
2. **Optional `root: ParentNode = document`** param on `initX()` / `initFlowbite()`. No behavior change at existing call sites.
3. **`AutoInit` module** — single `MutationObserver` on `document.body`, teardown-then-init on `childList` mutations. Rewrites `index.turbo.ts` to drop the four event listeners. Framework-agnostic — also helps Phoenix LiveView, Next.js, Nuxt, Blazor, Angular SSR.

Closes/relates to: #796, #1042, #1055, #998, #1102, #1111, #1123, #986, #1051, #1027, #989.

**Downstream (separate gem I'm building):** a Stimulus-controller library that honors Flowbite's existing `data-*-toggle` markup conventions, so any Flowbite copy-paste example works in a Hotwire app with proper lifecycle. This is happening regardless of upstream decisions — it gets way better if the upstream PRs land, because the Stimulus controllers can lean on `AutoInit` instead of re-implementing it.

PR 1 is already implemented on a fork (`mwlang/flowbite#fix/idempotent-init`); I'll smoke-test against Rails 8 + Turbo Drive/Frame/Stream before pushing.

**Three yes/no questions:**

1. **Will you merge this approach if PR 1 lands clean?** If you have a different direction in mind for #796, please sketch it and I'll redirect. The work happens either way; the question is whether it happens upstream or on a published fork.
2. **`flowbite.turbo.js` backward compat:** is removing the four event listeners acceptable in a v4.x minor (PR 3 makes them no-ops anyway), or do you want them kept for one release behind an opt-in flag and removed in v5?
3. **New dev dep — Vitest + JSDOM** for idempotency / AutoInit lifecycle unit tests. Yes or no?

**Default action if I don't hear back within ~2 weeks:** I open PR 1 as-is (safe regardless of #2 and #3), continue PRs 2 and 3, and ship the downstream Stimulus library. I'd genuinely rather coordinate with you than ship in parallel — let me know.

**Comment by pldavid2** [comment]
@mwlang thanks for this! Indeed it would be great to have this officially supported but being the repository was almost in maintenance mode for several months now, I suspect the whole AI wave is taking owners perhaps in a different direction.
One of the last commits was a link to typeui, a library that has several releases lately.
@zoltanszogyenyi any particular suggestion about the proposed changes and future roadmap of flowbite itself?
Thanks for your work
